diff --git a/Security/CasListener.php b/Security/CasListener.php
index cc93fd3..b045d90 100644
--- a/Security/CasListener.php
+++ b/Security/CasListener.php
@@ -49,7 +49,12 @@ class CasListener implements ListenerInterface {
 			$force = false;
 			if(!isset($_SESSION['cas_user'])) {
 				$auth = \phpCAS::checkAuthentication();
-				if($auth) $_SESSION['cas_user'] = \phpCAS::getUser();
+				if($auth) {
+                    $_SESSION['cas_user'] = \phpCAS::getUser();
+                    if (\phpCAS::hasAttributes()) {
+                        $_SESSION['cas_attributes'] = \phpCAS::getAttributes();
+                    }
+                }
 				else $_SESSION['cas_user'] = false;
 			}
 		}
@@ -62,7 +67,11 @@ class CasListener implements ListenerInterface {
 				$token = new CasToken();
 				$token->setUser($_SESSION['cas_user']);
 			}
-			$this->tokenStorage->setToken($this->authenticationManager->authenticate($token));
+            $authenticatedToken = $this->authenticationManager->authenticate($token);
+            if (isset($_SESSION['cas_attributes'])) {
+                $authenticatedToken->setAttributes($_SESSION['cas_attributes']);
+            }
+			$this->tokenStorage->setToken($authenticatedToken);
 			return;
 		}
 
